README.md

```diff
@@ -476,3 +509,55 @@ Finally, digital signature helps us use Content Delivery Network without
 worrying that a CDN compromise would end-up serving bad data to our users.
 Signing at the source reduces the pressure off of the infrastructure and
 allows us to rely on vendors without worrying about data integrity.
+
+### Algorithm
+
+## Addon Signing (XPI)
+
+### Algorithm
+
+Addons are XPI files (zip archives) where each file in the archives is hashed
+with SHA-1 and MD5, and the manifest containing all the hashes is signed. To
+request a signature, the client must first compute the manifest file (stored in
+`META-INF/manifest.mf` in the XPI), then make a signature file (stored in
+`META-INF/mozilla.sf`) and submit the `sf` file to the signing API, alongside
+the addon ID (typically an email address). The signing API returns an S/MIME
+detached signature the client must store under `META-INF/mozilla.rsa`.
+
+The signature file `mozilla.sf` simply contains md5.sha1 hashes of `manifest.mf`.
+(eg. `openssl dgst -binary -sha1 manifest.mf | base64`)
+
+To compute the signature, the signing API first generates an RSA key pair for
+the addon, called "ephemeral" because thrown away after signature. The certificate
+containing the public ephemeral key is signed by an intermediate issuer, itself
+signed by the AMO Root CA trusted in Firefox.
+
+The addon's certificate must have the following properties
+* SHA-256 signed
+* RSA public key of 2048 bits
+* subject CN set to the addon identifier (addon_id in the POST request)
+* 10 years validity
```

Comment:
Just a heads up, we once had a chemspill bug around expirations that weren't set right. I can't find it in bugzilla right now though.